front: highlight unindentified prs in itinerary component

Caracol3 opened the pull request:
close #8652

emersion commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Destination.tsx:
```diff
@@ -8,16 +9,23 @@ import { useOsrdConfActions, useOsrdConfSelectors } from 'common/osrdContext';
+  const isInvalid =
+    invalidPathItems &&
+    destination &&
+    'trigram' in destination &&
+    invalidPathItems.includes(destination.trigram);
```
Could we avoid using the `trigram` to identify path items? Path items all have an ID (note, _not_ the OP ID, which is stored separately in `operational_point`).

emersion commented:
This will need rebasing on top of https://github.com/OpenRailAssociation/osrd/pull/9289 (sorry!).

clarani commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Origin.tsx:
```diff
@@ -12,9 +13,10 @@ import { makeEnumBooleans } from 'utils/constants';
 
 type OriginProps = {
   zoomToFeaturePoint: (lngLat?: Position, id?: string) => void;
+  invalidPathItems?: string[];
```
I think it would be more elegant to give isInvalid to each component (Origin, Via and Destination) instead of the whole array invalidPathItems. 
You can create a utils function to check whether a pathItem is invalid or not and then call it in Itinerary.
```suggestion
  isInvalid?: boolean;
```

emersion replied:
I think we don't have `PathStep`s objects at this point. If one `PathItemLocation` is invalid, we don't have a pathfinding result. We only have `TrainScheduleBase.paths` and nothing else IIRC.

emersion replied:
The above seems to be false: the frontend strips invalid `PathItemLocation` entries and re-laucnhes pathfinding. So in the end we do have `PathStep`s.

Related: https://github.com/OpenRailAssociation/osrd/pull/9270#discussion_r1806605475

clarani replied:
It's good for me, @emersion ?

emersion replied:
Yeah, the new `isInvalid` field sgtm :)

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/ItineraryMarkers.tsx:
```diff
@@ -73,7 +78,7 @@ const extractMarkerInformation = (pathSteps: (PathStep | null)[], showStdcmAsset
           marker: cur,
           imageSource: showStdcmAssets ? stdcmVia : viaSVG,
-          index,
+          index: index,
```
```suggestion
          index,
```

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -177,9 +177,24 @@ export const usePathfinding = (
+
+    const filteredPathSteps = pathSteps.filter(
+      (step) =>
+        step !== null &&
+        step.coordinates !== null &&
+        !('trigram' in step && invalidItems.includes(step.trigram))
```
I think this is enough :)
```suggestion
        pathStepIsInvalid(step, invalidItems)
```

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/ItineraryMarkers.tsx:
```diff
@@ -57,9 +58,13 @@ const formatPointWithNoName = (
+  invalidItems: string[] = []
+) => {
+  return pathSteps.reduce((acc, cur, index) => {
+    if (cur && cur.coordinates && (!('trigram' in cur) || !invalidItems.includes(cur.trigram))) {
```
This condition is enough: if you don't find the pathItem in the infra, you won't have any coordinates. You can remove invalidItems from this component
```suggestion
    if (cur && cur.coordinates) {
```

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Vias.tsx:
```diff
@@ -37,7 +45,7 @@ const Vias = ({ zoomToFeaturePoint, shouldManageStopDuration }: ViasProps) => {
         {(provided) => (
           <div {...provided.droppableProps} ref={provided.innerRef}>
-            {vias.map((via, index) => (
+            {pathSteps.map((via, index) => (
```
Why do you use pathSteps instead of vias here ?

Caracol3 replied:
I struggled yesterday with the Vias component so had to do with the PathSteps but indeed, today (with a clearer mind 😁) there is no problem using vias.

clarani commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Vias.tsx:
```diff
@@ -37,7 +45,7 @@ const Vias = ({ zoomToFeaturePoint, shouldManageStopDuration }: ViasProps) => {
         {(provided) => (
           <div {...provided.droppableProps} ref={provided.innerRef}>
-            {vias.map((via, index) => (
+            {pathSteps.map((via, index) => (
```
Why do you use pathSteps instead of vias here ?

clarani commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Vias.tsx:
```diff
@@ -12,13 +12,21 @@ import { useAppDispatch } from 'store';
+const Vias = ({
+  zoomToFeaturePoint,
+  shouldManageStopDuration,
+  pathSteps,
```
No need to give pathSteps in this component, you can directly use the vias (from the selector getVias)

Caracol3 replied:
done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/ItineraryMarkers.tsx:
```diff
@@ -57,9 +57,9 @@ const formatPointWithNoName = (
-  pathSteps.reduce((acc, cur, index) => {
-    if (cur && cur.coordinates) {
+const extractMarkerInformation = (pathSteps: (PathStep | null)[], showStdcmAssets: boolean) => {
+  return pathSteps.reduce((acc, cur, index) => {
```
You can remove the return here and the brackets ;)

Caracol3 replied:
done

clarani commented on front/src/styles/scss/common/components/_pathfinding.scss:
```diff
@@ -91,6 +91,12 @@
+.via,
+.place {
+  &.invalid-path-item {
+    color: #e05206;
```
Do you think you can use a color from the theme ? You can use theme('colors.orange') for instance (if Thibaut defined an orange color), or directly var(--orange)

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -225,8 +236,24 @@ export const usePathfinding = (
+            if (invalidPathItems.length > 0) {
+              const updatedPathSteps = pathSteps.map((step) => {
+                if (step && 'trigram' in step && invalidPathItems.includes(step.trigram)) {
+                  return { ...step, invalid: true };
```
Is it invalid or isInvalid ?

Caracol3 replied:
My bad. It is indeed 'isInvalid'.

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -225,8 +236,24 @@ export const usePathfinding = (
+          if (pathfindingResult.status === 'invalid_path_items') {
+            const invalidPathItems = pathfindingResult.items
+              .map((item) => {
+                return 'trigram' in item.path_item ? item.path_item.trigram : null;
```
Can you add a comment TODO here to precise that we currently handle only invalid pathSteps with trigram ? Later we will have to handle invalid pathSteps with trackOffset, opId and uic

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -225,8 +236,24 @@ export const usePathfinding = (
+              });
+              dispatch(updatePathSteps({ pathSteps: updatedPathSteps }));
+              pathfindingDispatch({ type: 'PATHFINDING_PARAM_CHANGED' });
+            }
```
Maybe you can dispatch here a message for the user which explains that the pathfinding has failed because some pathSteps were not found on the infra ?
We should display this message even if invalidPathSteps is empty (as it contains only the invalidPathSteps with trigram and not all the invalidPathSteps)

Caracol3 replied:
Now even if there is an invalid pathSteps, the pathfinding succeed (if there is at least two valid pathSteps).
Do we still want to dispatch a message then ?

clarani reviewed (commented):
Super nice work :) almost there

clarani commented on front/src/modules/pathfinding/components/Itinerary/Itinerary.tsx:
```diff
@@ -49,6 +50,8 @@ const Itinerary = ({
   const { t } = useTranslation('operationalStudies/manageTrainSchedule');
   const { openModal } = useModal();
 
+  const isPathStepInvalid = (step: PathStep | null): boolean => step?.isInvalid || false;
```
Do you think you can move this function in a utils ?

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/components/Itinerary/DisplayItinerary/Vias.tsx:
```diff
@@ -16,9 +17,10 @@ import ViaStopDurationSelector from './ViaStopDurationSelector';
 type ViasProps = {
   shouldManageStopDuration?: boolean;
   zoomToFeaturePoint: (lngLat?: Position, id?: string) => void;
+  isInvalid: (step: PathStep | null) => boolean;
```
Since this function is in an utils, you can remove this props and directly use the function

Caracol3 replied:
done

clarani commented on front/package.json:
Can you make this change in a separated commit ? You can call me if you don't know how to do this :)

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -230,7 +241,38 @@ export const usePathfinding = (
             pathfindingResult.error_type === 'incompatible_constraints';
 
-          if (pathfindingResult.status === 'success' || incompatibleConstraintsCheck) {
+          const isInvalidPathItems =
```
Maybe it's a better name
```suggestion
          const hasInvalidPathItems =
```

clarani reviewed (commented):
LGTM :white_check_mark:

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -230,7 +241,38 @@ export const usePathfinding = (
+            pathfindingResult.error_type === 'invalid_path_items';
+
+          if (isInvalidPathItems) {
+            const invalidPathItems = pathfindingResult.items
+              .map((item) => ('trigram' in item.path_item ? item.path_item.trigram : null))
+              .filter((trigram): trigram is string => trigram !== null);
+
+            if (invalidPathItems.length > 0) {
+              const updatedPathSteps = pathSteps.map((step) => {
+                if (step && 'trigram' in step && invalidPathItems.includes(step.trigram)) {
+                  return { ...step, isInvalid: true };
+                }
+                return step;
+              });
+
+              dispatch(updatePathSteps({ pathSteps: updatedPathSteps }));
+              pathfindingDispatch({ type: 'PATHFINDING_FINISHED' });
+              pathfindingDispatch({ type: 'PATHFINDING_PARAM_CHANGED' });
+            } else {
+              dispatch(
+                setFailure(
+                  castErrorToFailure(pathfindingResult, {
+                    name: t('pathfinding_failed'),
+                    message: t('missing_pathSteps'),
+                  })
+                )
+              );
+            }
```
Maybe you can move this in a function handleInvalidPathItems ?

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -208,6 +221,51 @@ export const usePathfinding = (
     }
   }, [origin?.id, destination?.id, rollingStock]);
 
+  const handleInvalidPathItems = (
+    pathfindingResult: PathfindingResult,
+    pathSteps: (PathStep | null)[],
+    dispatch: AppDispatch,
+    pathfindingDispatch: React.Dispatch<PathfindingAction>,
+    t: TFunction
+  ) => {
```
As you are in the component, you can simplify the arguments of your function:
```suggestion
  const handleInvalidPathItems = (
    pathSteps: (PathStep | null)[],
    invalidPathItems: Extract<PathfindingInputError, { error_type: 'invalid_path_items' }>['items']
  ) => {
```

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -229,8 +287,13 @@ export const usePathfinding = (
+            pathfindingResult.failed_status === 'pathfinding_input_error' &&
+            pathfindingResult.error_type === 'invalid_path_items'
+          ) {
+            handleInvalidPathItems(pathfindingResult, pathSteps, dispatch, pathfindingDispatch, t);
```
Don't give the pathfindingResult, but directly the invalid pathSteps: `pathfindingResult.items`
```suggestion
            handleInvalidPathItems(pathSteps, pathfindingResult.items);
```

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -208,6 +221,51 @@ export const usePathfinding = (
+    pathfindingDispatch: React.Dispatch<PathfindingAction>,
+    t: TFunction
+  ) => {
+    if (
+      pathfindingResult.status === 'failure' &&
+      pathfindingResult.failed_status === 'pathfinding_input_error' &&
+      pathfindingResult.error_type === 'invalid_path_items'
+    ) {
```
Since you have the invalidItems in the arguments of your function, you can remove this if
```suggestion
```

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -208,6 +221,51 @@ export const usePathfinding = (
+      pathfindingResult.error_type === 'invalid_path_items'
+    ) {
+      //TODO: we currently only handle invalid pathSteps with trigram. We will have to do it for trackOffset, opId and uic too.
+      const invalidPathItems = pathfindingResult.items
```
You can rename this props
```suggestion
      const invalidTrigrams = pathfindingResult.items
```

Caracol3 replied:
done

clarani commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -208,6 +221,51 @@ export const usePathfinding = (
+        pathfindingDispatch({ type: 'PATHFINDING_PARAM_CHANGED' });
+      } else {
+        dispatch(
+          setFailure(
+            castErrorToFailure(pathfindingResult, {
+              name: t('pathfinding_failed'),
+              message: t('missing_pathSteps'),
+            })
+          )
```
You can directly setFailure with a name and a message :)
```suggestion
         setFailure({
            name: t('pathfinding_failed'),
            message: t('missing_pathSteps'),
          })
```

Caracol3 replied:
done

clarani reviewed (commented):
You can squash and rebase :)

clarani commented on front/public/locales/en/operationalStudies/manageTrainSchedule.json:
```diff
@@ -79,6 +79,7 @@
   "launchPathFinding": "Launch pathfinding",
   "launchSimulation": "Start",
   "manageVias": "Steps management",
+  "missing_pathSteps": "At least one pathstep wasn't found in the infrastructure",
```
We usually use CamelCase, even though I find snake case more readable, especially for translations :/
```suggestion
  "missingPathSteps": "At least one pathstep wasn't found in the infrastructure",
```

Caracol3 replied:
done

clarani reviewed (commented):
LGTM :white_check_mark: Can you rebase and clean your commits ? :) I will approve once you have cleaned it

emersion commented on front/package.json:
```diff
@@ -151,6 +151,9 @@
     "vitest": "^2.0.0"
   },
+  "optionalDependencies": {
+    "@rollup/rollup-linux-arm64-gnu": "^4.23.0"
```
Could this be split into its own commit? It's unrelated to the bug this PR fixes.

emersion replied:
Oh, I just noticed @clarani has made the same request above.

Small note: your new commit is missing the "front:" prefix.

emersion commented on front/src/modules/pathfinding/hooks/usePathfinding.ts:
```diff
@@ -256,9 +302,11 @@ export const usePathfinding = (
                 const correspondingOp = suggestedOperationalPoints.find(
                   (suggestedOp) =>
-                    'uic' in step && suggestedOp.uic === step.uic && suggestedOp.ch === step.ch
+                    ('uic' in step && suggestedOp.uic === step.uic && suggestedOp.ch === step.ch) ||
+                    (suggestedOp.trigram === trigramValue && suggestedOp.ch === step.ch)
```
Instead of adding more cases in this function, could we use `matchPathStepAndOp()`? That function already handles all of the possible cases.

Caracol3 replied:
Indeed ! Much clearer with the function. Thanks ;-)

Caracol3 committed:
front: highlight unindentify prs in itinerary component

- adding isInvalid to the PathSteps type
- moving isPathStepInvalid function in a utils file

Signed-off-by: Mathieu <mathieu.coulibaly@sncf.fr>

Caracol3 committed:
chore: update package.json with optional dependencies for mac users

Signed-off-by: Mathieu <mathieu.coulibaly@sncf.fr>

clarani reviewed (approved):
LGTM :white_check_mark: Nice job :clap:
